vm/src/symboltable.rs

```diff
@@ -0,0 +1,272 @@
+/* Python code is pre-scanned for symbols in the ast.
+
+This ensures that global and nonlocal keywords are picked up.
+Then the compiler can use the symbol table to generate proper
+load and store instructions for names.
+*/
+
+use rustpython_parser::ast;
+use std::collections::HashMap;
+
+pub enum SymbolRole {
+    Global,
+    Nonlocal,
+    Used,
+    Assigned,
+}
+
+pub struct SymbolTable {
+    // TODO: split-up into nested scopes.
+    symbols: HashMap<String, SymbolRole>,
+}
+
+impl SymbolTable {
+    pub fn new() -> Self {
+        SymbolTable {
+            symbols: HashMap::new(),
+        }
+    }
+
+    pub fn lookup(&self, name: &str) -> Option<&SymbolRole> {
+        return self.symbols.get(name);
+    }
+
+    pub fn scan_program(&mut self, program: &ast::Program) {
+        self.scan_statements(&program.statements);
+    }
+
+    pub fn scan_statements(&mut self, statements: &[ast::LocatedStatement]) {
+        for statement in statements {
+            self.scan_statement(statement)
+        }
+    }
+
+    fn scan_statement(&mut self, statement: &ast::LocatedStatement) {
+        match &statement.node {
+            ast::Statement::Global { names } => {
+                for name in names {
+                    self.register_name(name, SymbolRole::Global);
+                }
+            }
+            ast::Statement::Nonlocal { names } => {
+                for name in names {
+                    self.register_name(name, SymbolRole::Nonlocal);
+                }
+            }
+            ast::Statement::FunctionDef {
+                name,
+                body,
+                args,
+                decorator_list,
+                returns,
+            } => {
+                self.scan_expressions(decorator_list);
+                self.register_name(name, SymbolRole::Assigned);
+                for parameter in &args.args {}
```

Comment:
Shouldn't these names be marked assigned?

Reply:
I guess so, yes.